README
This is Perl module Plack::App::GAMobile.

INSTALLATION

Plack::App::GAMobile installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Plack::App::GAMobile

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Plack::App::GAMobile documentation is available as in POD. So you can do:

    % perldoc Plack::App::GAMobile

to read the documentation online with your favorite pager.

Masahiro Nagano


Changes
Revision history for Perl extension Plack::App::GAMobile

0.01    Wed Mar  2 12:14:55 2011
        - original version


lib/Plack/GAMobile.pm
package Plack::GAMobile;

use strict;
use warnings;
use Carp;
use Digest::MD5;
use URI;
use URI::QueryParam;

use Plack::Util::Accessor qw/base_url account secret/;

sub new {
    my $class = shift;
    my $self;
    if (@_ == 1 && ref $_[0] eq 'HASH') {
        $self = bless {%{$_[0]}}, $class;
    } else {
        $self = bless {@_}, $class;
    }

    Carp::croak "google analytics mobile beacon base_url is needed" unless $self->base_url;
    Carp::croak "google analytics account id is needed" unless $self->account;
    Carp::croak "sercret key for checksum is needed" unless $self->secret;

    $self;
}

sub image_url {
    my $self = shift;
    my $env = shift;

    my $utmn = int(rand 0x7fffffff );
    my $referer = defined $env->{HTTP_REFERER} ? $env->{HTTP_REFERER} : "-";
    my $path = defined $env->{REQUEST_URI} ? $env->{REQUEST_URI} : "";

    my $domain = "";
    if ( defined $env->{HTTP_HOST} ) {
        $domain = $env->{HTTP_HOST};
    }
    elsif ( defined $env->{SERVER_NAME} ) {
        $domain = $env->{SERVER_NAME};
    }

    #guid cleanup
    if ( $referer ne "-") {
        my $r_uri = URI->new($referer);
        $r_uri->query_param_delete('guid');
        $referer = $r_uri->as_string;
    }

    my $digest = substr( Digest::MD5::md5_hex($self->secret . $utmn . $domain . $path), 16, 6 );

    my $url = URI->new($self->base_url);
    $url->query_form_hash({
        utmac => $self->account,
        utmn  => $utmn,
        utmhn => $domain,
        utmr  => $referer,
        utmp  => $path,
        cs    => $digest,
        guid  => 'ON',
    });
    return $url;
}

1;

__END__

=head1 NAME

Plack::GAMobile - Google Analytics Mobile Beacon Image URL generator

=head1 SYNOPSIS

use Plack::GAMobile;

  my $gam = Plack::GAMobile->new(
      secret => 'my very secret key',
      base_url => '/gam',
      account => 'ACCOUNT ID GOES HERE',
  );

  my $image_url = $gam->image_url($env);

  # in template
  <img src="[% image_url %]" />

=head1 DESCRIPTION

Plack::GAMobile is Plack::App::GAMobile's client. This module generate beacon image url.

=head1 METHOD

=over 4

=item new

Create instance of Plack::GAMobile.

=over 4

=item base_url

Base URL of beacon image.

=item secret

Secret key of making checksum. Set same secret of Plack::App::GAMobile.

=item account

Account ID of your Google Analytics ID.

=back

=item image_url($env)

generate beacon image url. $env is PSGI env.

=back

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo {at} gmail.comE<gt>

=head1 SEE ALSO

L<Plack::App::GAMobile>, L<http://code.google.com/intl/ja/mobile/analytics/>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/WWW/GoogleAnalytics/Mobile/PSGI.pm
package WWW::GoogleAnalytics::Mobile::PSGI;

use strict;
use warnings;
use parent qw/Plack::Component/;
use Carp;
use URI;
use URI::QueryParam;
use List::Util qw/first/;
use Furl;
use Net::DNS::Lite;
use Digest::MD5;
use Plack::Request;
use Plack::Request;

use Plack::Util::Accessor qw/secret timeout/;

our $GAM_VERSION = '4.4sp';
our $GAM_COOKIE_NAME = '__utmmobile';
our $GAM_COOKIE_PATH = '/';
our $GAM_COOKIE_USER_PERSISTENCE = '+2y';
our $GAM_UTM_GIF_LOCATION = "http://www.google-analytics.com/__utm.gif";

my $GIF_DATA = pack "C43", (
    0x47, 0x49, 0x46, 0x38, 0x39, 0x61,
    0x01, 0x00, 0x01, 0x00, 0x80, 0xff,
    0x00, 0xc0, 0xc0, 0xc0, 0x00, 0x00,
    0x00, 0x21, 0xf9, 0x04, 0x01, 0x00,
    0x00, 0x00, 0x00, 0x2c, 0x00, 0x00,
    0x00, 0x00, 0x01, 0x00, 0x01, 0x00,
    0x00, 0x02, 0x02, 0x44, 0x01, 0x00,
    0x3b );

sub prepare_app {
    my $self = shift;
    Carp::croak "secret key must be defined" if ! $self->secret;
    $self->timeout(5) unless defined $self->timeout;
}

sub param_or {
    my ($req, $key, $val) = @_;
    return $req->param($key) if defined $req->param($key);
    return $val;
}

sub call {
    my $self = shift;
    my $env = shift;
    my $req = Plack::Request->new($env);

    my $domain_name = param_or($req, 'utmhn', "");
    my $document_referer = param_or($req, 'utmr', "-");
    my $document_path = param_or($req, 'utmp', "");
    my $account = param_or($req, 'utmac', "");
    my $utmn = param_or($req, 'utmn', "");

    if ( ! $req->param('cs') ) {
        return ['403', ['Content-Type' => 'text/plain'], ['no checksum'] ];
    }
    if ( $req->param('cs') ne
             substr( Digest::MD5::md5_hex($self->secret . $utmn . $domain_name . $document_path), 16, 6 ) ) {
        return ['403', ['Content-Type' => 'text/plain'], ['checksum not match'] ];
    }

    my $user_agent = "";
    if ( defined $req->user_agent ) {
        $user_agent = $req->user_agent;
    }

    my $remote_address = "";
    if ( defined $req->address ) {
        $remote_address = $req->address;
        if ($remote_address =~ /^((\d{1,3}\.){3})\d{1,3}$/) {
            $remote_address = $1 . "0";
        } else {
            $remote_address = "";
        }
    }

    srand(); #for preforking
    my $visitor_id = $self->get_visitor_id($req, $user_agent, $account);
    
    my $utm_url = URI->new($GAM_UTM_GIF_LOCATION);
    $utm_url->query_form_hash({
        utmwv  => $GAM_VERSION,
        utmn   => int(rand 0x7fffffff),
        utmhn  => $domain_name,
        utmr   => $document_referer,
        utmp   => $document_path,
        utmac  => $account,
        utmcc  => '__utma=999.999.999.999.999.1;',
        utmvid => $visitor_id,
        utmip  => $remote_address
    });

    my @headers;
    if (defined $req->header("Accept-Language") ) {
        push @headers, "Accept-Language", $req->header("Accept-Language");
    }
    
    my $furl = Furl::HTTP->new(
        inet_aton => \&Net::DNS::Lite::inet_aton,
        timeout   => $self->timeout,
        agent     => $user_agent,
        headers   => \@headers,
    );
    $furl->env_proxy;
    my ($minor_version, $status, $message, $headers, $content) = $furl->get("$utm_url");

    if ( substr( $status, 0, 1 ) ne '2' ) {
        Carp::carp "Failed request to '$GAM_UTM_GIF_LOCATION': $message";
    }

    my $res = Plack::Response->new(200);
    $res->content_type('image/gif');
    $res->header('Cache-Control', 'private, no-cache, no-cache=Set-Cookie, proxy-revalidate');
    $res->header('Pragma', 'no-cahce');
    $res->cookies->{$GAM_COOKIE_NAME} = {
        value => $visitor_id,
        path => $GAM_COOKIE_PATH,
        expires => $GAM_COOKIE_USER_PERSISTENCE,
    };
    $res->body($GIF_DATA);

    return $res->finalize;
}

sub get_visitor_id {
    my ($self, $req, $user_agent, $account) = @_;

    my $guid = first { defined $_ } map { $req->env->{$_}  } (
        "HTTP_X_DCMGUID",
        "HTTP_X_UP_SUBNO",
        "HTTP_X_JPHONE_UID",
        "HTTP_X_EM_UID"
    );
    $guid = "" if ! defined $guid;

    my $cookie = "";
    if ( defined $req->cookies->{$GAM_COOKIE_NAME} ) {
        $cookie = $req->cookies->{$GAM_COOKIE_NAME};
    }

    return $cookie if ($cookie ne "");

    my $message = "";
    if ($guid ne "") {
        $message = $guid . $account;
    } else {
        $message = $user_agent . int(rand 0x7fffffff );
    }

    my $md5_string = Digest::MD5::md5_hex($message);
    return "0x" . substr($md5_string, 0, 16);
}


1;
__END__

=head1 NAME

GoogleAnalytics::Mobile::PSGI - Server-side PSGI application of Google Analytics for Mobile

=head1 SYNOPSIS

  use WWW::GoogleAnalytics::Mobile::PSGI;
  use Plack::Builder;

  builder {
      mount "/gam" => WWW::GoogleAnalytics::Mobile::PSGI->new(
          secret => 'my very secret key',
          timeout => 4,
      );
  };

=head1 DESCRIPTION

The server-side PSGI application of Google Analytics Mobile.

=head1 METHOD

=over 4

=item new

=over 4

=item secret

Secret key of checksum. Set same secret of WWW::GoogleAnalytics::Mobile

=item timeout

Timeout second of request to Google.

=back

=back

=head1 OPTIMIZE

=over 4

=item proxy

Set environment value to use http proxy for accessing to Google. 

  my $app = WWW::GoogleAnalytics::Mobile::PSGI->new()->to_app;
  $app = sub {
      local $ENV{http_proxy} = 'http://172.0.0.3';
      $app->(shift);
  };

=item Cache DNS queries

WWW::GoogleAnalytics::Mobile::PSGI uses L<Furl::HTTP> and L<Net::DNS::Lite> for requesting to Google. You can cache DNS queries with Cache::LRU for more speed.

  use WWW::GoogleAnalytics::Mobile::PSGI;
  use Cache::LRU;

  # setup cache for Net::DNS::Lite
  $Net::DNS::Lite::CACHE = Cache::LRU->new(
      size => 256,
  );

=back

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo {at} gmail.comE<gt>

=head1 SEE ALSO

L<WWW::GoogleAnalytics::Mobile>, L<http://code.google.com/intl/ja/mobile/analytics/>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/WWW/GoogleAnalytics/Mobile.pm
package WWW::GoogleAnalytics::Mobile;

use strict;
use warnings;
use Carp;
use Digest::MD5;
use URI;
use URI::QueryParam;

use Plack::Util::Accessor qw/base_url account secret/;

our $VERSION = '0.01';

sub new {
    my $class = shift;
    my $self;
    if (@_ == 1 && ref $_[0] eq 'HASH') {
        $self = bless {%{$_[0]}}, $class;
    } else {
        $self = bless {@_}, $class;
    }

    Carp::croak "google analytics mobile beacon base_url is needed" unless $self->base_url;
    Carp::croak "google analytics account id is needed" unless $self->account;
    Carp::croak "sercret key for checksum is needed" unless $self->secret;

    $self;
}

sub image_url {
    my $self = shift;
    my $env = shift;

    my $utmn = int(rand 0x7fffffff );
    my $referer = defined $env->{HTTP_REFERER} ? $env->{HTTP_REFERER} : "-";
    my $path = defined $env->{REQUEST_URI} ? $env->{REQUEST_URI} : "";

    my $domain = "";
    if ( defined $env->{HTTP_HOST} ) {
        $domain = $env->{HTTP_HOST};
    }
    elsif ( defined $env->{SERVER_NAME} ) {
        $domain = $env->{SERVER_NAME};
    }

    #guid cleanup
    if ( $referer ne "-") {
        my $r_uri = URI->new($referer);
        $r_uri->query_param_delete('guid');
        $referer = $r_uri->as_string;
    }

    my $digest = substr( Digest::MD5::md5_hex($self->secret . $utmn . $domain . $path), 16, 6 );

    my $url = URI->new($self->base_url);
    $url->query_form_hash({
        utmac => $self->account,
        utmn  => $utmn,
        utmhn => $domain,
        utmr  => $referer,
        utmp  => $path,
        cs    => $digest,
        guid  => 'ON',
    });
    return $url;
}

1;

__END__

=head1 NAME

WWW::GoogleAnalytics::Mobile - Image URL generator of Google Analytics for Mobile beacon

=head1 SYNOPSIS

  use WWW::GoogleAnalytics::Mobile;

  my $gam = WWW::GoogleAnalytics::Mobile->new(
      secret => 'my very secret key',
      base_url => '/gam',
      account => 'ACCOUNT ID GOES HERE',
  );

  my $image_url = $gam->image_url($env);

  # in template
  <img src="[% image_url %]" />

  # server-side
  use WWW::GoogleAnalytics::Mobile::PSGI;
  use Plack::Builder;

  builder {
      mount "/gam" => WWW::GoogleAnalytics::Mobile::PSGI->new(
          secret => 'my very secret key',
          timeout => 4,
      );
      $app;
  };



=head1 DESCRIPTION

WWW::GoogleAnalytics::Mobile makes URI of Google Analytics of Mobile beacon that runs
by WWW::GoogleAnalytics::Mobile::PSGI.

=head1 METHOD

=over 4

=item new

Create instance of WWW::GoogleAnalytics::Mobile

=over 4

=item base_url

Base URL of beacon image.

=item secret

Secret key of making checksum. Set same secret of WWW::GoogleAnalytics::Mobile::PSGI

=item account

Account ID of your Google Analytics ID.

=back

=item image_url($env)

generate beacon image url. $env is PSGI env.

=back

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo {at} gmail.comE<gt>

=head1 SEE ALSO

L<WWW::GoogleAnalytics::Mobile::PSGI>, L<http://code.google.com/intl/ja/mobile/analytics/>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

